### DISCUSSION TITLE
Proposal: Keep a stack of current z-index value to automatically set z-index on popovers and avoid `withinPortal: false`

### DISCUSSION THREAD

**Discussion by KurtGokhan** [mantinedev/mantine#6554] [Q&A]
## Problem

To render some components like Select inside popover, [the docs](https://mantine.dev/core/select/#inside-popover) suggest to use `withinPortal: false`. There are also other workarounds like setting z-index.

The problem with `withinPortal: false` is, it is often more useful to render within a Portal than not, because of overflow issues. So, often the only valid option is to use z-index in possibly every place a component with a portal is used.

This breaks composability. If I have a complex component, I want it to work well outside popovers, as well as inside popovers, modals or drawers. We shouldn't have to tweak the z-index depending on where we are rendering the component.

I want to make a remark that this is not a Mantine specific problem. Z-index and stacking order is the most complicated thing in CSS in my opinion. There is no absolute way to tame z-index without trade-offs. The good thing is, Mantine has a system that is perfectly suited to handle most common issues.

## Suggested Solution

Let's say we have a numerical CSS variable named `--mantine-current-z-index` that is 0 by default. Every time a Popover is opened, it should get this variable from its target, add it with the default or user-provided z-index, and pass it to the Popover.Dropdown. This way, every children of the dropdown will also inherit this CSS variable in its incremented state, and popovers opened inside it will be opened with a higher z-index.

The calculated z-index can be made reactive with a property like `zIndexDependencies` similar to `positionDependencies` does for calculated position. Alternatively `positionDependencies` can be reused for the same purpose to recalculate z-index.

For Drawers and Modals (and other kind of non-targeted overlays), `--mantine-current-z-index` should be set to the z-index of its root element. If a mechanism to implement stacked modals/drawers is implemented in the future by any chance, this can be tweaked and even be helpful in implementing such a mechanism.

I should note that this works best if z-index is configured for different elements in exponentially growing values. If for example popover has default z-index of 10 and drawer has 20, three-times nested popovers would have a z-index of 30 and that would be higher than the drawer. Instead the drawers should have like a z-index of 100 or 1000, as nesting that many popovers are unlikely.

Let me know what you think and if you need help/clarification.

**Comment by rtivital** [comment]
> To render some components like Select inside popover, the docs suggest to use withinPortal: false. There are also other workarounds like setting z-index.

It is not issue with z-index. It is related to click outside events. If the dropdown node is not part of the DOM tree of the popover it will be closed when one of the dropdown options is clicked.

> This breaks composability. If I have a complex component, I want it to work well outside popovers, as well as inside popovers, modals or drawers. We shouldn't have to tweak the z-index depending on where we are rendering the component.

By default, if you do not change z-index values, there aren't any issues with overlays – Mantine provides default values for z-index that are designed to work well.

**Reply by KurtGokhan** [reply]
For reasons, we had to mix different design systems and we had issues probably nobody else will have. If you think the defaults are all that is needed, I am ok with that.

I never encountered the click-outside issue so I assumed the issue was about z-index, since it also solves that. Maybe the docs should mention the reason it is recommended.

By the way, without portals, you can have problems like this if the dropdown hides the overflow:

![image](https://github.com/user-attachments/assets/af9d7ea0-7d98-4b8c-b5fd-bff63063e636)

Sometimes this can be solved with fixed floating strategy, but not when the dropdown creates a stacking context (e.g. animating). Do you know a solution to this?
